cvmfs/swissknife_list_reflog.cc

```diff
@@ -0,0 +1,177 @@
+    success = Run(&object_fetcher, repo_name, output_path, reflog_hash);
+  }
+
+  return (success) ? 0 : 1;
+}
+
+template <class ObjectFetcherT>
+bool CommandListReflog::Run(ObjectFetcherT *object_fetcher, string repo_name,
+                            string output_path, shash::Any reflog_hash)
+{
+  typename ObjectFetcherT::ReflogTN *reflog;
+  reflog = FetchReflog(object_fetcher, repo_name, reflog_hash);
+
+  // Traverse through catalogs and regular objects
+  vector<shash::Any> catalogs;
+  if (NULL == reflog || !reflog->List(SqlReflog::kRefCatalog, &catalogs)) {
+    LogCvmfs(kLogCvmfs, kLogStderr, "Failed to list catalog reference log");
+    return false;
+  }
+  typename CatalogTraversal<ObjectFetcherT>::Parameters traversal_params;
+  traversal_params.object_fetcher = object_fetcher;
+  CatalogTraversal<ObjectFetcherT> traversal(traversal_params);
+  traversal.RegisterListener(
+    &swissknife::CommandListReflog::CatalogCallback,
+    this);
+  bool success = true;
+  vector<shash::Any>::iterator i = catalogs.begin();
+  const vector<shash::Any>::const_iterator iend = catalogs.end();
+  for (; i != iend && success; i++) {
+    success &= traversal.TraverseRevision(*i,
+                CatalogTraversal<ObjectFetcherT>::kBreadthFirstTraversal);
+  }
+
+  if (!success) {
+    LogCvmfs(kLogCvmfs, kLogStderr,
+             "Catalog traversal aborted due to an error");
+  }
+
+  // Add history, certificate, metainfo objects from reflog
+  vector<shash::Any> histories, certificates, metainfos;
+  if (!reflog->List(SqlReflog::kRefHistory, &histories)) {
+    LogCvmfs(kLogCvmfs, kLogStderr,
+             "Failed to fetch history objects from reflog");
+  }
+  if (!reflog->List(SqlReflog::kRefCertificate, &certificates)) {
+    LogCvmfs(kLogCvmfs, kLogStderr,
+             "Failed to fetch certificate objects from reflog");
+  }
+  if (!reflog->List(SqlReflog::kRefMetainfo, &metainfos)) {
+    LogCvmfs(kLogCvmfs, kLogStderr,
+             "Failed to fetch metainfo objects from reflog");
+  }
+  objects_.insert(histories.begin(), histories.end());
+  objects_.insert(certificates.begin(), certificates.end());
+  objects_.insert(metainfos.begin(), metainfos.end());
+
+  // Clean up reflog file
+  reflog->DropDatabaseFileOwnership();
+  const string reflog_db = reflog->database_file();
+  delete reflog;
+  unlink(reflog_db.c_str());
+
+  LogCvmfs(kLogCvmfs, kLogStdout, "Number of objects: %lu", objects_.size());
+
+  if (output_path != "") {
+    DumpObjectsToFile(output_path);
+  }
+
+  return success;
+}
+
+void CommandListReflog::CatalogCallback(
+  const CatalogTraversalData<catalog::Catalog> &data)
+{
+  const catalog::Catalog::HashVector &referenced_hashes =
+    data.catalog->GetReferencedObjects();
+  objects_.insert(referenced_hashes.begin(), referenced_hashes.end());
+  objects_.insert(data.catalog->hash());
+}
+
+void CommandListReflog::DumpObjectsToFile(const string &path)
```

Comment:
It might be easier to simply write string by string with `FILE *` and `fwrite`. The C I/O does buffering internally.